encfs/CipherFileIO.cpp

```diff
@@ -377,42 +535,127 @@ ssize_t CipherFileIO::writeOneBlock(const IORequest &req) {
+         * backing file has not been modified by others since we opened it
+         * (otherwise getSize() could return a wrong cached result). */
+        if ((padBytes > 1) || (req.offset + bs >= getSize())) {
+          // this is the last block, we must pad it with cipherBlockSize() bytes
+          memset(req.data + writeSize, 0x00, cbs - padBytes);
+          writeSize += cbs - padBytes;
+        }
+      }
+    } else {
+      // let's drop last bytes at a cipherBlockSize() modulo,
+      // they could be padding or result from a partial write
+      int padOutBytes = writeSize % cbs;
+      writeSize -= padOutBytes;
+      if (writeSize > 0) {
+        ok = blockWrite(req.data, (int)writeSize,
+                        blockNum ^ fileIV);  // cast works because we work on a
+                                             // block and blocksize fit an int
+        if (ok) {
+          int padInBytes = 0;
+          // let's look for the 0x00 padding bytes at the end inside the data block
+          while ((padInBytes < (cbs - padOutBytes - 1)) && (req.data[writeSize - padInBytes - 1] == 0x00)) {
+            padInBytes++;
+          }
+          // let's look for the first and needed padding byte, 0x80
+          if (req.data[writeSize - padInBytes - 1] == 0x80) {
+            padInBytes++;
+          } else {
+            padInBytes = 0;
+          }
+          // let's check if fully-sized block is correctly padded
+          if (writeSize == bs) {
+            if ((padInBytes >= 1) && (padInBytes <= cbs)) {
+              writeSize -= padInBytes;
+            } else {
+              VLOG(1) << "writeOneBlock failed (wrong padding) for block " << blockNum << ", size "
+                      << writeSize;
+              writeSize = -EBADMSG;
+            }
+          // for a partial block (could be a partial write), don't write what looks like padding
+          } else if ((padOutBytes + padInBytes) <= cbs) {
+            writeSize -= padInBytes;
+          }
+        }
+      }
+    }
   }
 
-  ssize_t res = 0;
-  if (ok) {
-    if (haveHeader) {
-      IORequest tmpReq = req;
-      tmpReq.offset += HEADER_SIZE;
-      res = base->write(tmpReq);
+  // no CBC padding, stream cipher instead
+  else {
+    if (req.dataLen != bs) {
+      ok = streamWrite(req.data, (int)writeSize,
+                       blockNum ^ fileIV);  // cast works because we work on a
+                                            // block and blocksize fit an int
     } else {
-      res = base->write(req);
+      ok = blockWrite(req.data, (int)writeSize,
+                      blockNum ^ fileIV);  // cast works because we work on a
+                                           // block and blocksize fit an int
     }
-  } else {
+  }
+
+  if (!ok) {
     VLOG(1) << "encodeBlock failed for block " << blockNum << ", size "
             << req.dataLen;
-    res = -EBADMSG;
+    writeSize = -EBADMSG;
   }
-  return res;
+
+  if (writeSize > 0) {
```

Comment:
please check "writeSize <= 0" early and return out. Gets rid of one level of indentation.